File: frontend/src/models/bikes/bike.ts
import {DataArrayResponse, FetchResponse, getOrigin, httpDelete, httpGetWithBody, httpPost} from "@/models/httpMethods";
import {EquipmentType} from "@/models/equipmentTypes/equipmentType";

export interface IBike {
    id: string;
    manufacturer: string;
    model: string;
    mileage: number;
    typeId: string;

    addBikeRequest(): Promise<FetchResponse<null>>;

    deleteBikeRequest(): Promise<FetchResponse<null>>;

    clear(): void;
}

export class Bike implements IBike {
    id: string;
    manufacturer: string;
    model: string;
    mileage: number;
    typeId: string;


    async addBikeRequest(): Promise<FetchResponse<null>> {
        return await httpPost<IBike>("/api/bike", this);
    }

    async deleteBikeRequest(): Promise<FetchResponse<null>> {
        return await httpDelete("/api/bike", this.id);
    }

    clear(): void {
        this.manufacturer = "";
        this.model = "";
        this.mileage = 0.0;
        this.typeId = "";
    }

    // Copies over the fields in the argument object if given
    constructor(...args: IBike[]) {
        this.id = "";
        this.manufacturer = "";
        this.model = "";
        this.mileage = 0.0;
        this.typeId = "";
        if (args.length === 1) {
            Object.assign(this, args[0]);
        }
    }
}
export async function getBikesRequest(): Promise<IBike[]> {
    let result = await httpGetWithBody<DataArrayResponse<Bike>>("/api/bike");
    if (result.status === 200) {
        return result.body.data.map(createBike);
    }
    return [];
}


function createBike(bike: IBike) {
    return new Bike(bike);
}

File: BikeHistory/Models/Bikes/Bike.cs
namespace BikeHistory.Models.Bikes;

public class Bike
{
    public Guid Id { get; set; }
    public string Manufacturer { get; set; }
    public string Model { get; set; }
    public double Mileage { get; set; }
    public Guid TypeId { get; set; }
    public List<ServiceEntry> Services {get;set;}

    public BikeDto CreateDto()
    {
        return new BikeDto(Id, Manufacturer, Model, Mileage, TypeId);
    }

    public Bike(string manufacturer, string model, double mileage)
    {
        Id = Guid.NewGuid();
        Manufacturer = manufacturer;
        Model = model;
        Mileage = mileage;
        Services = new List<ServiceEntry>();
    }
}
